engines/watchmaker/2d_stuff.cpp

```diff
@@ -0,0 +1,187 @@
+	int32 a = 0;
+	for (a = 0; a < MAX_DD_BITMAPS; a++)
+		if (!DDBitmapsList[a].tnum)
+			return a;
+
+	warning("Too many DD Bitmaps!");
+	return -1;
+}
+
+void TwoDeeStuff::displayDDBitmap(int32 tnum, int32 px, int32 py, int32 ox, int32 oy, int32 dx, int32 dy) {
+	int32 a = findFreeBitmap();
+	if (a == -1) {
+		warning("Skipping draw");
+		return;
+	}
+
+	DDBitmapsList[a].tnum = tnum;
+	DDBitmapsList[a].px   = _renderer->rFitX(px);
+	DDBitmapsList[a].py   = _renderer->rFitY(py);
+	DDBitmapsList[a].ox   = _renderer->rFitX(px + ox) - _renderer->rFitX(px);
+	DDBitmapsList[a].oy   = _renderer->rFitY(py + oy) - _renderer->rFitY(py);
+	DDBitmapsList[a].dx   = _renderer->rFitX(px + dx) - _renderer->rFitX(px);
+	DDBitmapsList[a].dy   = _renderer->rFitY(py + dy) - _renderer->rFitY(py);
+	if (dx <= 0) DDBitmapsList[a].dx += _renderer->getBitmapDimX(tnum) - DDBitmapsList[a].ox;
+	if (dy <= 0) DDBitmapsList[a].dy += _renderer->getBitmapDimY(tnum) - DDBitmapsList[a].oy;
+}
+
+void TwoDeeStuff::displayDDBitmap_NoFit(int32 tnum, int32 px, int32 py, int32 ox, int32 oy, int32 dx, int32 dy) {
+	int32 a = findFreeBitmap();
+	if (a == -1) {
+		warning("Skipping draw");
+		return;
+	}
+
+	DDBitmapsList[a].tnum = tnum;
+	DDBitmapsList[a].px   = (px);
+	DDBitmapsList[a].py   = (py);
+	DDBitmapsList[a].ox   = (px + ox) - (px);
+	DDBitmapsList[a].oy   = (py + oy) - (py);
+	DDBitmapsList[a].dx   = (px + dx) - (px);
+	DDBitmapsList[a].dy   = (py + dy) - (py);
+	if (dx <= 0) DDBitmapsList[a].dx += _renderer->getBitmapDimX(tnum) - DDBitmapsList[a].ox;
+	if (dy <= 0) DDBitmapsList[a].dy += _renderer->getBitmapDimY(tnum) - DDBitmapsList[a].oy;
+}
+
+int32 TwoDeeStuff::rendDDText(char *text, FontKind font, FontColor color) {
+	struct SDDText *r;
+	int32 c, tdx, tdy;
+	char info[100];
+
+	if ((!text) || (text[0] == '\0')) return -1;
+
+	for (c = 0, r = &RendText[0]; c < MAX_REND_TEXTS; c++, r++) {
+		if (r->text[0]) continue;
+		// Get the size of the text to render
+		_renderer->_fonts->getTextDim(text, font, &tdx, &tdy);
+		// Create a surface to contain it
+		r->tnum = rCreateSurface(tdx, tdy, rBITMAPSURFACE);
+		_renderer->clearBitmap(r->tnum, 0, 0, tdx, tdy, 0, 0, 0);
+		// Render the lettering on the surface
+		//DebugLogWindow("Creo testo %s | %d %d",text,tdx,tdy );
+		_renderer->printText(text, r->tnum,  font, color, 0, 0);
+		Common::strlcpy(info, "text: ", sizeof(info));
+		strncat(info, text, 15);
+		//DebugLogWindow("Creato %s",info);
+		rSetBitmapName(r->tnum, info);
+		Common::strlcpy(r->text, text, sizeof(r->text));
+		r->color = color;
+		r->font = font;
+		return r->tnum;
+	}
+
+	return -1;
+}
+
+void TwoDeeStuff::displayDDText(char *text, FontKind font, FontColor color, int32 px, int32 py, int32 ox, int32 oy, int32 dx, int32 dy) {
+	struct SDDText *t, *r;
+	int32 a, c;
+
+	if ((!text) || (text[0] == '\0')) return;
+
+	for (a = 0; a < MAX_DD_TEXTS; a++)
+		if (!DDTextsList[a].text[0])
+			break;
+
+	if (a >= MAX_DD_TEXTS) {
+		warning("Too many DD Texts!");
```

Comment:
We already have an exclamation mark at the end of our warning. Also, I'd recommend prefixing such warnings with the method name, e.g. `"TwoDeeStuff::displayDDText(): Too many DD Texts"`